### PR TITLE
[RCTIPlus] Support events and TV

### PR DESCRIPTION
**Pull request by MinePlayersPE**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

Adds support for live/missed events, and TV (live events but with their own section + they presumably run forever)
Also fixes regex matching for series, adds basic geoblock detection, and fixes `live_status` docs (as per @pukkandan's request)

### CODE REVIEW

File: yt_dlp/extractor/rcti.py [hunk 1]
```diff
@@ -30,7 +32,7 @@ def _call_api(self, url, video_id, note=None):
 
 
 class RCTIPlusIE(RCTIPlusBaseIE):
-    _VALID_URL = r'https://www\.rctiplus\.com/programs/\d+?/.*?/(?P<type>episode|clip|extra)/(?P<id>\d+)/(?P<display_id>[^/?#&]+)'
+    _VALID_URL = r'https://www\.rctiplus\.com/(programs/\d+?/.*?/)?(?P<type>episode|clip|extra|live-event|missed-event)/(?P<id>\d+)/(?P<display_id>[^/?#&]+)'
```
Comment (pukkandan) [inline yt_dlp/extractor/rcti.py:35]:
```suggestion
    _VALID_URL = r'https://www\.rctiplus\.com/(?:programs/\d+?/.*?/)?(?P<type>episode|clip|extra|live-event|missed-event)/(?P<id>\d+)/(?P<display_id>[^/?#&]+)'
```

File: yt_dlp/extractor/rcti.py [hunk 2]
```diff
@@ -142,35 +191,44 @@ def _real_extract(self, url):
...
                 'url': '%s%d%s' % (image_path, 2000, video_meta['landscape_image'])
             })
-
-        formats = self._extract_m3u8_formats(video_url, display_id, 'mp4', headers={'Referer': 'https://www.rctiplus.com/'})
+        try:
+            formats = self._extract_m3u8_formats(video_url, display_id, 'mp4', headers={'Referer': 'https://www.rctiplus.com/'})
+        except ExtractorError as e:
+            if isinstance(e.cause, compat_HTTPError) and e.cause.code == 403:
+                self.raise_geo_restricted(countries=['ID'], metadata_available=True)
+            raise e
```
Comment (pukkandan) [inline yt_dlp/extractor/rcti.py:199]:
```suggestion
            else:
                raise e
```

File: yt_dlp/extractor/rcti.py [hunk 3]
```diff
@@ -142,35 +191,44 @@ def _real_extract(self, url):
...
+            if 'akamaized' in f['url'] or 'cloudfront' in f['url']:
+                f.setdefault('http_headers', {})['Referer'] = 'https://www.rctiplus.com/'  # Referer header is required for akamai/cloudfront CDNs
 
         self._sort_formats(formats)
 
         return {
             'id': video_meta.get('product_id') or video_json.get('product_id'),
-            'title': video_meta.get('title') or video_json.get('content_name'),
+            'title': (video_meta.get('title') or video_meta.get('name')
+                      or video_json.get('content_name') or video_json.get('assets_name')),
```
Comment (pukkandan) [inline yt_dlp/extractor/rcti.py:209]:
```suggestion
            'title': dict_get(video_meta, ('title', 'name')) or dict_get(video_json, ('content_name', 'assets_name')),
```

File: yt_dlp/extractor/rcti.py [hunk 4]
```diff
@@ -142,35 +191,44 @@ def _real_extract(self, url):
...
             'season_number': video_meta.get('season'),
             'episode_number': video_meta.get('episode'),
             'channel': video_json.get('tv_name'),
             'channel_id': video_json.get('tv_id'),
             'formats': formats,
-            'thumbnails': thumbnails
+            'thumbnails': thumbnails,
+            'is_live': video_type == 'live-event' and not is_upcoming,
+            'was_live': video_type == 'missed-event',
+            'live_status': 'is_upcoming' if is_upcoming else None,
```
Comment (pukkandan) [inline yt_dlp/extractor/rcti.py:225]: You can also just set `live_status` and the other two will be automatically set

File: yt_dlp/extractor/rcti.py [hunk 5]
```diff
@@ -253,3 +315,48 @@ def _real_extract(self, url):
...
+
+    def _real_extract(self, url):
+        match = re.match(self._VALID_URL, url).groupdict()
+        tv_id = match.get('tvname') or match.get('eventname')
+        webpage = self._download_webpage(url, tv_id)
+        event_re = re.search(
+            r'url ?: ?"https://api\.rctiplus\.com/api/v./(?P<type>[^/]+)/(?P<id>\d+)/url', webpage)
+        if not event_re:
+            raise RegexNotFoundError('Unable to extract video link')
+        video_type, video_id = event_re.groups()
```
Comment (pukkandan) [inline yt_dlp/extractor/rcti.py:357]: self._search_regex

File: yt_dlp/extractor/rcti.py [hunk 6]
```diff
@@ -253,3 +315,48 @@ def _real_extract(self, url):
...
+        return {
+            '_type': 'url',
+            'url': 'https://www.rctiplus.com/%s/%s/%s' % (video_type, video_id, tv_id),
+            'ie_key': 'RCTIPlus',
+        }
```
Comment (pukkandan) [inline yt_dlp/extractor/rcti.py]: self.url_result

File: yt_dlp/extractor/common.py [hunk 7]
```diff
@@ -297,7 +297,7 @@ class InfoExtractor(object):
                     live stream that goes on instead of a fixed-length video.
     was_live:       True, False, or None (=unknown). Whether this video was
                     originally a live stream.
-    live_status:    'is_live', 'upcoming', 'was_live', 'not_live' or None (=unknown)
+    live_status:    'is_live', 'is_upcoming', 'was_live', 'not_live' or None (=unknown)
```
Comment (pukkandan) [inline yt_dlp/extractor/common.py:300]: I believe this is broken in readme too

File: yt_dlp/extractor/rcti.py [hunk 8]
```diff
@@ -253,3 +316,44 @@ def _real_extract(self, url):
...
+    def _real_extract(self, url):
+        match = re.match(self._VALID_URL, url).groupdict()
+        tv_id = match.get('tvname') or match.get('eventname')
+        webpage = self._download_webpage(url, tv_id)
+        event_re = re.search(
+            r'url ?: ?"https://api\.rctiplus\.com/api/v./(?P<type>[^/]+)/(?P<id>\d+)/url', webpage)
+        if not event_re:
+            raise RegexNotFoundError('Unable to extract video link')
+        video_type, video_id = event_re.groups()
+        return self.url_result('https://www.rctiplus.com/%s/%s/%s' % (video_type, video_id, tv_id), 'RCTIPlus')
```
Comment (pukkandan) [inline yt_dlp/extractor/rcti.py:359]: btw, u can use fstrings to simplify this and the api callls if u want

File: yt_dlp/extractor/rcti.py [hunk 9]
```diff
@@ -351,9 +350,6 @@ def _real_extract(self, url):
         match = re.match(self._VALID_URL, url).groupdict()
         tv_id = match.get('tvname') or match.get('eventname')
         webpage = self._download_webpage(url, tv_id)
-        event_re = re.search(
-            r'url ?: ?"https://api\.rctiplus\.com/api/v./(?P<type>[^/]+)/(?P<id>\d+)/url', webpage)
-        if not event_re:
-            raise RegexNotFoundError('Unable to extract video link')
-        video_type, video_id = event_re.groups()
+        video_type, video_id = self._search_regex(
+            r'url ?: ?"https://api\.rctiplus\.com/api/v./(?P<type>[^/]+)/(?P<id>\d+)/url', webpage, 'video link', group=('type', 'id'))
```
Comment (pukkandan) [inline yt_dlp/extractor/rcti.py:354]:
```suggestion
            r'url\s*:\s*["\']https://api\.rctiplus\.com/api/v./(?P<type>[^/]+)/(?P<id>\d+)/url', webpage, 'video link', group=('type', 'id'))
```

a bit more robust

File: yt_dlp/extractor/rcti.py (hunk 8 above)
Comment (MinePlayersPE) [reply yt_dlp/extractor/rcti.py:355]: done for long api calls, though i don't think i'd want to replace everything

Comment (MinePlayersPE) [reply yt_dlp/extractor/rcti.py:355]: done for long api calls, though i don't think i'd want to replace everything

Comment (pukkandan) [reply yt_dlp/extractor/rcti.py:355]: sure, just letting you know its an option

File: yt_dlp/extractor/rcti.py (hunk 4 above)
Comment (MinePlayersPE) [reply yt_dlp/extractor/rcti.py:225]: sure, but it honestly seems simpler this way

**Pull request merged**